File: sth/public/js/controllers/budget_controller.js
// Copyright (c) 2025, DAS and Contributors
// MIT License. See license.txt

sth.plantation.month = [
    'januari', 'februari', 'maret', 'april', 'mei', 'juni',
    'juli', 'agustus', 'september', 'oktober', 'november', 'desember'
];

sth.plantation.setup_budget_controller = function() {
    sth.plantation.BudgetController = class BudgetController extends sth.plantation.TransactionController {
        setup(doc) {
            super.setup(doc)

            let me = this
            for (const month of sth.plantation.month) {
                frappe.ui.form.on(doc.doctype, `per_${month}`, function() {
                    me.calculate_total_sebaran()
                });
            }
        }
        
        set_query_field(){
            super.set_query_field()
        
            this.frm.set_query("budget_kebun_tahunan", function(doc){
                if(!doc.company){
                    frappe.throw("Please Select Company First")
                }

                return{
                    filters: {
                        company: doc.company
                    }
                }
            })

        }

        company(){
            this.frm.set_value("budget_kebun_tahunan", "")
        }
 
        rotasi(_, cdt, cdn){
            this.calculate_total(cdt, cdn)
        }
        
        is_distibute(doc){
            if(doc.is_distibute){

                for (const month of sth.plantation.month) {
                    this.frm.doc[`per_${month}`] = 100 / 12
                }
                
                this.calculate_total_sebaran()
            }
        }
        
        calculate_total_sebaran(){
            let me = this
            let total_sebaran = 0.0
            
            // update sebaran di semua table
            me.doctype_ref("table_fieldname").forEach(table_ref => {
                let per_month_table = Object.keys(
                    me.frm.fields_dict[table_ref].grid.fields_map
                ).filter(key => key.includes("rp_"));
                // jika tidak ada fieldname degan kata per_ skip sebaran

                if(per_month_table.length == 0) return

                for (const item of me.frm.doc[table_ref] || []) {
                    me.calculate_sebaran_values(item, per_month_table)
                }
            });

            
            // set nilai sebaran
            for (const month of sth.plantation.month) {
                total_sebaran += flt(this.frm.doc[`per_${month}`] || 0)
            }

            this.frm.doc.total_sebaran = total_sebaran
            
            me.frm.refresh_fields();
        }

        update_value_after_amount(item){
           item.amount = flt(item.amount * (item.rotasi || 1), precision("amount", item));
        }
        
        calculate_sebaran_values(item, months=[]){
            // set nilai sebaran
            for (const month of months) {
                let per_month = month.replace(/^rp_/, "per_")
                
                item[month] = flt(
                    item.amount * (this.frm.doc[per_month] / 100),
                    precision(month, item)
                );
            }
        }

        add_blok_in_table(blok_table, args, default_field={}){
            let me = this

            this.get_blok_list({ args }, (data) => {
                let cur_grid = this.frm.fields_dict[blok_table].grid;
                data.forEach(blok => {
                    let row = frappe.model.add_child(this.frm.doc, cur_grid.doctype, blok_table);
                    let ignore_fields = [
                        "name",
                        "idx",
                        "__checked"
                    ];

                    for (let key in blok) {
                        if (in_list(ignore_fields, key)) {
                            continue;
                        }

                        row[key] = blok[key];
                    }
                    
                    if (default_field){
                        $.each(default_field, (key, value) => {
                            row[key] = value
                        })
                    }

                });

                refresh_field(blok_table);
                me.frm.get_field(blok_table).tab.set_active();
            })
        }

        after_calculate_item_values(table_name, total){
            // update nilai sebaran dan rotasi jika ada
            let data_table = this.frm.doc[table_name] || []

            // set on child class if 
            let per_month_table = Object.keys(
                this.frm.fields_dict[table_name].grid.fields_map
            ).filter(key => key.includes("rp_"));

            for (const item of data_table) {
                // jika tidak ada fieldname degan kata per_ skip sebaran    
                if(per_month_table.length == 0) continue
                this.calculate_sebaran_values(item, per_month_table)
            }
        }

        clear_table(list_table=[]){
            for (const field_table of list_table || []) {
                this.frm.clear_table(field_table)
                this.calculate_item_values(field_table)
            }

            this.calculate_grand_total();
            this.frm.refresh_fields();
        }
    }
}

File: sth/controllers/budget_controller.py
# Copyright (c) 2025, DAS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, get_link_to_form

from sth.controllers.plantation_controller import PlantationController

class BudgetController(PlantationController):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.duplicate_param = ["budget_kebun_tahunan"]

    def validate(self):
        self.check_duplicate_data()
        super().validate()

        self.check_total_sebaran()
        self.calculate_item_sebaran_and_rotasi()

    def check_duplicate_data(self):
        if not self.duplicate_param:
            return
        
        filters = {"docstatus": 1, "name": ["!=", self.name]}
        for param in self.duplicate_param:
            filters[param] = self.get(param)

        if doc := frappe.db.get_value(self.doctype, filters):
            frappe.throw("{} is already use in <b>{}</b>".format(self.doctype, get_link_to_form(self.doctype, doc)))

    def update_value_after_amount(self, item, precision):
        # set on child class if needed
        item.amount = flt(item.amount * (item.get("rotasi") or 1), precision)

    def calculate_item_sebaran_and_rotasi(self):
        for df in self._get_table_fields():        
            rotasi_total = 0
            table_item = self.get(df.fieldname)

            # cari fieldname dengan kata rp pada table
            per_month_table = list(filter(lambda key: "rp_" in key, frappe.get_meta(df.options).get_valid_columns()))
            for d in table_item:
                rotasi_total += d.get("rotasi") or 0

                if per_month_table:
                    self.calculate_sebaran_values(d, per_month_table)

            # check jika ada rotasi untuk input total rotasi
            fieldname = f"{df.fieldname}_rotasi"
            if not self.meta.has_field(fieldname):
                continue

            self.set(fieldname, flt((rotasi_total / len(table_item)) if table_item else 0, self.precision(fieldname)))

    def calculate_sebaran_values(self, item, sebaran_list=[]):
        # hitung nilai sebaran selama 12 bulan
        for sbr in sebaran_list:
            per_field = sbr.replace("rp_", "per_")

            # default 0 frappe selalu berbentuk string
            per_month = flt(self.get(per_field))
            item.set(sbr, 
                flt(item.amount * (per_month / 100), item.precision(sbr))
            )

    def check_total_sebaran(self):
        total_sebaran = 0.0
        per_month_field = list(filter(lambda key: key.startswith("per_"), self.meta.get_valid_columns()))
        if not per_month_field:
            return 
            
        for month in per_month_field:
            if self.is_distibute:
                self.set(month, 100 / 12)

            total_sebaran += self.get(month)

        self.total_sebaran = flt(total_sebaran, self.precision("total_sebaran"))
        if self.total_sebaran != 100:
            frappe.throw(_(f"Total distribution is {'below' if self.total_sebaran < 100 else 'over'} 100%."))

    